synapse/util/caches/lrucache.py

```diff
@@ -82,19 +93,243 @@ def enumerate_leaves(node, depth):
+        self.next_node = None
+        self.prev_node = None
+
+        self.cache_entry = None
+
+    def move_after(self, root: "_ListNode", clock: Clock):
+        """Move this node from its current location in the list to after the
+        given node.
+        """
+        # We assert that both this node and the root node is still "alive".
+        assert self.prev_node
+        assert self.next_node
+        assert root.prev_node
+        assert root.next_node
+
+        # Remove self from the list
+        prev_node = self.prev_node
+        next_node = self.next_node
+
+        prev_node.next_node = next_node
+        next_node.prev_node = prev_node
+
+        # Insert self back into the list, after root
+        prev_node = root
+        next_node = root.next_node
+
+        self.prev_node = prev_node
+        self.next_node = next_node
+
+        prev_node.next_node = self
+        next_node.prev_node = self
+
+    def get_cache_entry(self) -> Optional[P]:
+        """Get the cache entry, returns None if this is the root node (i.e.
+        cache_entry is None) or if the entry has been dropped.
+        """
+        return self.cache_entry
+
+
+class _TimedListNode(_ListNode[P]):
+    """A `_ListNode` that tracks last access time."""
+
+    __slots__ = ["last_access_ts_secs"]
+
+    def __init__(self, clock: Clock, cache_entry: Optional[P]) -> None:
+        super().__init__(cache_entry=cache_entry)
+
+        self.last_access_ts_secs = int(clock.time())
+
+    @staticmethod
+    def insert_after(
+        cache_entry: P, root: "_ListNode[P]", clock: Clock
+    ) -> "_TimedListNode[P]":
+        node = _TimedListNode(clock, cache_entry)
+        node.move_after(root, clock)
+        return node
+
+    def move_after(self, root: "_ListNode", clock: Clock):
+        self.last_access_ts_secs = int(clock.time())
+        return super().move_after(root, clock)
+
+
+# Whether to insert new cache entries to the global list. We only add to it if
+# time based eviction is enabled.
+USE_GLOBAL_LIST = False
+
+# A linked list of all cache entries, allowing efficient time based eviction.
+GLOBAL_ROOT = _ListNode[_CacheEntry]()
+
+
+@wrap_as_background_process("LruCache._expire_old_entries")
+async def _expire_old_entries(clock: Clock, expiry_seconds: int):
+    """Walks the global cache list to find cache entries that haven't been
+    accessed in the given number of seconds.
+    """
+
+    now = int(clock.time())
+    node = GLOBAL_ROOT.prev_node
+    assert node is not None
+
+    i = 0
+    orphaned_nodes = 0
+
+    logger.debug("Searching for stale caches")
+
+    while node is not GLOBAL_ROOT:
+        # Only the root node isn't a `_TimedListNode`.
+        assert isinstance(node, _TimedListNode)
+
+        if node.last_access_ts_secs > now - expiry_seconds:
+            break
+
+        cache_entry = node.get_cache_entry()
+        current_node = node
+        node = node.prev_node
```

Comment:
these are confusing names. I suggest:
```suggestion
        next_node = node.prev_node
```

... and then do `node=next_node` at the bottom of the loop.